docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+
+# stopping a job:
+$ jobserver stop <job-id>
+<job-id>
+
+# getting the status of a job:
+$ jobserver status <job-id>
+state: running
+pid: 1234
+start time: 2006-01-02T15:04:05Z07:00
+
+# getting the output of a job:
+$ jobserver logs [--follow] <job-id>
+<output>
+
+# getting a list of all job ids:
+$ jobserver list
+JOB ID      COMMAND       CREATED          STATUS
+<id-1>      kubectl       2006-01-02...    Running
+<id-2>      go            2006-01-02...    Completed
+```
+
+### Implementation Details
+
+#### Streaming Output
+
+Output streaming will be implemented as follows:
+
+1. When a job is started, we will pipe the combined stdout and stderr of its process to a simple in-memory buffer.
+2. When a client initiates a stream by calling `Output()`, we will write the current contents of the buffer to the stream in 4MB chunks. Then, if the client requested to follow the output, the server will keep the stream open, and any subsequent reads from the process's output will be duplicated to all clients that are following the stream, in addition to the server's own buffer.
+
+#### Process Lifecycle
+
+For jobs that do not specify resource limits, they will be run as a child process of the job server itself without any additional isolation. For jobs that do specify resource limits, they will be run in their own cgroup with all of the requested limits.
+
+On startup, the server will do the following (simplifying, there will be additional abstractions here):
+
+1. Ensure its euid is 0 so it can manage cgroups, and ensure that cgroups v2 is enabled.
+2. Create a parent cgroup in `/user.slice/user-<uid>.slice/jobserver` if it doesn't exist. The specific path can be determined by first identifying its own cgroup from `/proc/self/cgroup` (which, if run as `sudo jobserver serve`, will be something like `/user.slice/user-<uid>.slice/<terminal emulator's cgroup>/vte-spawn-<some uuid>.scope`), then walking the tree up until it finds a cgroup matching `user-*.slice`.
+3. Enable the `cpu`, `memory`, and `io` controllers in the `jobserver` cgroup's `cgroup.subtree_control` file, if necessary.
+
+The lifecycle of a job's cgroup will be as follows:
+
+1. When a job is created with resource limits, the server will create a new cgroup with the name `<job-id>` under the parent `jobserver` cgroup.
+2. The server will write the requested resource limits to the new cgroup's `cpu.max`, `memory.max`, and `io.max` files (for whichever limits were specified).
+3. The server will create a file descriptor to the new cgroup by opening its path (`/sys/fs/cgroup/user.slice/user-<uid>.slice/jobserver/<job-id>`).
+4. The server will configure the `exec.Cmd`'s `SysProcAttr` with `CgroupFd` set to the file descriptor. When this is set, Go will call `clone3` with `CLONE_INTO_CGROUP` so that the new child process is started directly in the new cgroup.
+5. When the process exits, the server will close the file descriptor and delete the job's cgroup.
```

Comment:
So the "job" is tied to the continued existence of the first child, right? Could you elaborate on what is the plan to deal with other processes in the cgroup and/or processes that have inherited the same stdio pipes (and thus are holding them open)?

(subprocesses that deliberately escape the cgroup are out of scope for the challenge)

Reply:
Ah, good point - when the first child exits, we'll need to clean up any other processes in the cgroup since the cgroups won't have their own pid namespaces. So step 5 could instead be:

> 5. When the process exits, the server will, in order:
	  - close the file descriptor
	  - kill any remaining processes in the job's cgroup by writing "1" to its `cgroup.kill` file
	  - remove the job's cgroup directory

Would that work?

Reply:
Yeah, that looks correct! 👍